keps/sig-autoscaling/draft-20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,277 @@
+   for the given Pod.
+1. The initating actor observes that ResourceRequirements and
+   ResourceAllocated match again which signifies the completion of an update.
+
+#### Transitions of InPlaceResize condition
+
+The following diagram shows possible transitions of InPlaceResize condition.
+
+```text
+                   +---------+
+       +-----------+         +-----------+
+       |           | (empty) |           |
+       | +--------->         <---------+ |
+       | |         +----+----+         | |
+      1| |2            3|             4| |5
+ +-----v-+--+           |          +---+-v--+
+ |          |           |          |        |
+ | Awaiting |           |          | Failed |
+ |          |           |          |        |
+ +-------+--+           |          +---+----+
+        3|              |              |3
+         |         +----v-----+        |
+         |         |          |        |
+         +---------> Rejected <--------+
+                   |          |
+                   +----------+
+```
+
+1. Scheduler, on pre-emption.
+1. Scheduler, after pre-emption finishes.
+1. Any Controller, on permanent issue.
+1. Kubelet, on successful retry.
+1. Kubelet, if not enough space on Node.
+
+#### Notes
+
+* In case when there is no pre-emption required, Kubelet and Scheduler
+  will pick up the ResourceRequirements change in parallel.
+* In case when there is pre-emption required Kubelet and Scheduler might
+  pick up the ResourceRequirements change in parallel,
+  Kubelet will then set the InPlaceResize condition to Failed
+  and Scheduler will clear it once pre-emption is done.
+* Kubelet might try to apply new resources also if InPlaceResize
+  condition is set to Failed, as a normal retry mechanism.
+* To avoid races and possible gamification, all components should use
+  max(ResourceRequirements, ResourceAllocated) when computing resources
+  used by a Pod. TBD if this can be weakened when InPlaceResize condition
+  is set to Rejected, or should the initiating actor update
+  ResourceRequirements back to reclaim resources.
+
+### Affected Components
+
+Pod v1 core API:
+* extended model,
+* added validation.
+
+Admission Controllers: LimitRanger, ResourceQuota need to support Pod Updates:
+* for ResourceQuota it should be enough to change podEvaluator.Handler
+  implementation to allow Pod updates; max(ResourceRequirements, ResourceAllocated)
+  should be used to be in line with current ResourceQuota behaviour
+  which blocks resources before they are used (e.g. for Pending Pods),
+* for LimitRanger TBD.
+
+Kubelet
+* support in-place resource management,
+* set ResourceRequirements on placing the Pod on Node.
+
+Scheduler:
+* update its caches with proper resources, depending on InPlaceResize condition.
+
+Other components:
+* check how the change of meaning of resource requests influence other kubernetes components.
+
+### Possible Extensions
+
+1. Allow resource limits to be updated too.
```

Comment:
I may have missed this earlier.  The kep describes requests and limits, but this implies differently.  What is this referencing?

Reply:
If I'm not mistaken, this refers to extension to VPA project. IIRC, today VPA supports updating requests, but not limits. @kgolab please correct me if that's not the case.